dimod/core/scoped.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2025 D-Wave
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+
+__all__ = ['Scoped']
+
+
+class Scoped(abc.ABC):
+    """Abstract base class for components that allocate scope-bound resources.
+
+    :class:`.Scoped` requires the concrete implementation to provide resources
+    clean-up via :meth:`~.Scoped.close` method. With that, context manager
+    protocol is automatically supported (resources are released on context exit).
+
+    The scoped resource should either be used from a `runtime context`_, e.g.:
+
+    >>> with SomeScopedSampler() as sampler:       # doctest: +SKIP
+    ...     sampler.sample(...)
+
+    or explicitly disposed with a call to the :meth:`.close` method.
+
+    .. _runtime context: https://docs.python.org/3/reference/datamodel.html#with-statement-context-managers
```

Comment:
Consider replacing a hard URL with :std:doc:`With Statement Context Managers <python:reference/datamodel>`

disadvantage: not the place in the page; advantage: less likely to go stale

Reply:
Given that `datamodel` page is huge, and URLs of Python docs never change, I'm prone to leaving the full URL.

Reply:
I'd be that way inclined myself